# JSON files output by Expansion Hunter

JSON files generated by Expansion Hunter contain information about sample
parameters (`SampleParameters` field) and analysis results summarized by
locus (`LocusResults` field). The locus results contain these fields
 * `AlleleCount` The expected number of alleles at the locus
 * `Coverage` Estimated read coverage at the locus
 * `FragmentLength` The fragment size estimated from read pairs fully contained in either the left or right flank of the repeat region
 * `LocusId` Locus identifier
 * `ReadLength` Mean read length at the locus
 * `Variants` Genotypes and other information describing each variant
   analyzed at the locus

The structure of records in the `Variant` field  is different for repeats and
small variants (insertions, deletions, and swaps).

## Repeat records

Repeat records contain the following fields.
* `VariantId` Unique variant identifier
* `RepeatUnit` Repeat unit in the reference orientation
* `VariantType` Always set to "Repeat"
* `VariantSubtype` Either "Repeat" or "RareRepeat"
* `Genotype` Repeat genotype given by the size of each repeat allele
* `GenotypeConfidenceInterval` Size confidence interval for each repeat allele
* `CountsOfSpanningReads` Summary of identified spanning reads given as an
   array with entries `(n, m)` where `n` is the number of repeat units spanned
   by the flanking read and `m` is the number of such reads
* `CountsOfFlankingReads` An analog of `CountsOfSpanningReads` for in-repeat
  reads
* `CountsOfInrepeatReads` An analog of `CountsOfSpanningReads` for spanning
  reads
* `ReferenceRegion` 0-based half-open reference coordinates of the repeat region
  (`chrom:start-end`)

## Small variant records

Records for small variants contain the following fields.
* `VariantId` Unique variant identifier
* `VariantType` Always set to "SmallVariant"
* `VariantSubtype` Either "Insertion", "Deletion", or "Swap" 
* `RepeatUnit` Repeat unit in the reference orientation
* `Genotype` Formed, as usual, from `0`s and `1`s corresponding to ref and alt
  alleles
* `CountOfAltReads` Number of reads supporting the alt allele
* `CountOfRefReads` Number of reads supporting the ref allele
* `ReferenceRegion` Reference region of the variant
